UI redesign: use modern tooltip style for remaining tooltip elements

Ddharmani3 opened the pull request:
<!-- Describe your pull request here.-->
change style of tooltip elements mentioned in issue #24311 to modern hotkey_hints tooltip style

Fixes: <!-- Issue link, or clear description.-->
#24311 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

* All private message:

![image](https://user-images.githubusercontent.com/64723994/221901746-ef210fac-ac38-4dbf-ab0a-a24bcad68c28.png)
![image](https://user-images.githubusercontent.com/64723994/221901607-4f851125-8df9-4989-8eb1-c08d2ce6f4f1.png)

* View user card:
![image](https://user-images.githubusercontent.com/64723994/219296022-6758ca36-88f1-4f61-b897-cb48537de103.png)

* Expand/condense Message:
![image](https://user-images.githubusercontent.com/64723994/219296467-71ea8b09-600f-470b-852a-d743e7cf4a69.png)

* Menu(setting menu icon):
![image](https://user-images.githubusercontent.com/64723994/221906013-a66487a8-666e-4805-a963-d09625782856.png)



some more elements advertising keyboard shortcut:

In left sidebar:

* All messages:
![image](https://user-images.githubusercontent.com/64723994/219298133-8999a66c-1589-486e-ac90-c7cd8668523e.png)

* Recent conversations:
![image](https://user-images.githubusercontent.com/64723994/219298296-a03c31dd-9c25-4855-9a9a-c08193f3ac75.png)

* Drafts:
![image](https://user-images.githubusercontent.com/64723994/219298408-655a3575-2db0-4229-bdc1-b8fd918a6aff.png)



3 elements in streams settings:

* Create new stream:
![image](https://user-images.githubusercontent.com/64723994/221924841-f431d000-d0e5-4efc-9779-563b43e75535.png)

* Toggle subscription:
![image](https://user-images.githubusercontent.com/64723994/221924922-2d33fde7-32ab-4bb9-809b-5305d684efdc.png)

* View stream:
![image](https://user-images.githubusercontent.com/64723994/221925022-87ee4d30-fe9c-456f-a00b-4b416fd2316d.png)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
@timabbott can you please review this PR?

Pranav2612000 commented:
@Ddharmani3 I already had a PR ready [ https://github.com/zulip/zulip/pull/24384 ] and was waiting for the other contributor to respond back before completing the task. It's best to read through the comments timeline, see if there is a PR open and wait for zulipbot to assign it to you before picking up a task.
Anyway, adding some comments regarding the commit messages

Pranav2612000 reviewed (changes requested):
Can you squash the commits so that the follow the zulip commit discipline ( https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html )

For example, you don't need a seperate commit to fix the frontend lint checks.
Also, I'm not sure but maybe each tooltip could have been a separate commit which makes it much easy to review. ( And it's better to err towards small commits, as the zulip documentation says )

Pranav2612000 commented on static/js/tippyjs.js:
```diff
@@ -449,20 +449,35 @@ export function initialize() {
         appendTo: () => document.body,
     });
+
+    delegate("body", {
```
Since the config for this block is the same as the one used for the block with `target: "#show_all_private_messages"`, what do you think about merging these two blocks together?

cc: @timabbott @sayamsamal

Ddharmani3 replied:
Thank you for reviewing my work and providing valuable feedback. I appreciate your attention to detail and your suggestions for improvement. I will definitely take your recommendations into consideration and make the necessary updates.

Ddharmani3 commented:
> @Ddharmani3 I already had a PR ready [ https://github.com/[/pull/24384](https://github.com/zulip/zulip/pull/24384) ] and was waiting for the other contributor to respond back before completing the task. It's best to read through the comments timeline, see if there is a PR open and wait for zulipbot to assign it to you before picking up a task. Anyway, adding some comments regarding the commit messages

@Pranav2612000 I have found 2 more elements to apply the modern tooltip style. can I continue working on this issue if you don't mind?

Ddharmani3 commented:
@timabbott , I've updated the tooltip style to a modern tippyjs tooltip for all the elements that advertise keyboard shortcuts. Can you please review my pull request?

alya commented:
Thanks! Could you please update the capitalization in your commit messages to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)?

alya commented:
Let's add 500 ms delays to all these tooltips; they are currently much too invasive.

alya commented:
Also, if we are going to update any tooltips for the shortcuts at the top of the left sidebar, we have to do migrate them all, including Mentions and Starred messages. Otherwise, the experience feels very inconsistent.

Ddharmani3 commented:
@alya I have updated commits messages according to [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html), added tooltips for mentions and starred messages for consistency, and also added 500ms delay to all the tooltips. These changes are ready for review now.

alya commented:
Thanks! The left sidebar tooltips still feel too invasive to me personally, but I'd be OK with testing this from a UX perspective. @amanagr would you be up for reviewing this PR?

Joelute commented on web/src/tippyjs.js:
```diff
@@ -448,19 +448,37 @@ export function initialize() {
+            ".message_expander",
+            ".message_condenser",
+            ".top_left_all_messages",
+            ".top_left_recent_topics",
+            ".top_left_mentions",
+            ".top_left_starred_messages",
+            ".top_left_drafts",
+            "#settings-dropdown",
+            ".create_stream_button",
+            ".toggle-subscription-tooltip",
+            "#preview-stream-button",
+        ],
         placement: "bottom",
-        content: $t({
-            defaultMessage: "All direct messages (P)",
-        }),
+        onShow(instance) {
+            const $elem = $(instance.reference);
+            const $template = $("#" + $elem.attr("data-tooltip-template-id"));
+            instance.setContent(parse_html($template.html()));
+        },
```
What's the reason for removing this and replacing it with a custom function? I'm pretty sure you could just continue using `content` rather than creating a custom function and creating an extra template in the Handlebars file.

Ddharmani3 replied:
The onShow function is used instead of content to display hotkey hints in the tooltip. This approach is consistent with that used in PR [#22746](https://github.com/zulip/zulip/pull/22746), which this [issue](https://github.com/zulip/zulip/issues/24311) builds upon. 
If I have misunderstood the issue you raised, please let me know.

Joelute replied:
Sorry, I was the one that misunderstood the issue, that's fine then.

Joelute commented:
I think overall looks good IMO, posted a comment where I felt the changes were unnecessary.

Ddharmani3 commented:
@alya @amanagr I feel that this PR is ready for integration review. Could you please take a look and provide your feedback?

alya commented:
As noted above, the next step here is for @amanagr to review this PR. Since it hasn't been very long, I'm sure this PR is in his queue -- there is no need to post reminders yet.

timabbott commented on web/src/tippyjs.js:
```diff
@@ -452,20 +452,38 @@ export function initialize() {
     });
 
     delegate("body", {
-        target: "#show_all_private_messages",
```
I don't think you need to use this sort of hardcoded list of elements. Our defaults already register the `get_tooltip_content` function for getting the tooltip content from `data-tooltip-template-id`. In fact, @amanagr can the `.message_control_button` `onShow` method be deleted in favor of using the default `data-tooltip-template-id`?

So the only thing you'd; be customizing is the delay; for that, how about we create a variant called `.tippy-zulip-delayed-tooltip` that is identical to `tippy-zulip-tooltip` except it adds delay `LONG_HOVER_DELAY` (which should replace the 500ms values you have in this PR)? Then we can just apply that class to these elements rather than having a fragile list of selectors in this file.

timabbott commented:
I ended up taking a look at this and suggested the rewrite above. My other comment is that after doing the refactor to introduce `tippy-zulip-delayed-tooltip` as suggested above, I'd like to have each commit just modify a single part of the app to use the new tooltip library; those should rebase cleanly with respect to each other (since you won't be stacking a list of CSS selectors in `tippy.js`) and it'll be easier to review that way, as one can merge individual sections as they're ready.

Ddharmani3 commented:
@timabbott, I have addressed all the feedback you provided and made the necessary changes. The PR is now ready for review

timabbott commented:
@Ddharmani3 can you address the merge conflicts?

Ddharmani3 commented:
> @Ddharmani3 can you address the merge conflicts?

@timabbott, I have resolved the merged conflicts and updated the commit [ui_redesign: Modern tippy tooltip for compose-send-button.](https://github.com/zulip/zulip/pull/24399/commits/7965e157c390818e0180bda62b33d815f63e5a33) to include the functionality of updating the tippy tooltip on the `compose-send-button` whenever the user's `enter send choice` changes. The original work for this approach was done in #24620 by @aparna-bhatt.

alya commented:
I'm not sure why the "has conflicts" label didn't get automatically removed.

Ddharmani3 commented:
> I'm not sure why the "has conflicts" label didn't get automatically removed.
> 
> 

I was thinking that too, maybe it's a glitch as I have resolved all the conflicts.

timabbott commented:
Thanks for your work on this @Ddharmani3, but I don't think I can integrate any of this as written. A few notes:
* The first commit contains changes that belong in later commits and thus is broken as written.
* The `.tippy-zulip-delayed-tooltip` class changes the tooltip position to `bottom`, which is a poor choice for the sidebar tooltips that are the majority of this PR, and also it's a very bad choice of name to have something that sounds like it just changes delay but also changes two other random configuration properties. I think we want the same left (fallback to right) behavior we have for the other tooltips in the left sidebar. I think realistically we need to instead make a `tippy-left-sidebar-tooltip` class that is more semantic that has the desired positioning and delays for the left sidebar (and would also unify with the existing behavior for private messages rows).
* For the ones that don't have a keyboard shortcut, like "Mentions" and "Starred messages", I think we'd be better off deleting the `title` attribute than adding a tooltip that has no new information. This would be a good first commit of a new PR that just works on the left sidebar.
* I'd prefer `tooltips:` as the commit message prefix, as described in our commit messages guidelines, the part before the `:` should be the subsystem being changed.

Closing this PR as I think it'd be better to start a fresh PR with these ideas in mind (without all the above history) than continue on this PR ID.
